arch/x86/x86_64/include/uk/asm/spinlock.h

```diff
@@ -0,0 +1,116 @@
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef __UKARCH_SPINLOCK_H__
+#error Do not include this header directly
+#endif
+
+#include <uk/arch/atomic.h>
+
+/* Unless you know what you are doing, use struct uk_spinlock instead. */
+typedef struct __spinlock {
+	volatile int lock;
+} __spinlock;
+
+/* Initialize a spinlock to unlocked state */
+#define UKARCH_SPINLOCK_INITIALIZER() { 0 }
+
+/**
+ * Initialize a spinlock to unlocked state.
+ * @param [in/out] lock Pointer to spinlock
+ */
+static inline void ukarch_spin_init(__spinlock *lock)
+{
+	lock->lock = 0;
+}
+
+/**
+ * Acquire lock. It is guaranteed that the spinlock will be held exclusively.
+ * @param [in/out] lock Pointer to spinlock
+ */
+static inline void ukarch_spin_lock(__spinlock *lock)
+{
+	register int locked = 1;
+
+	__asm__ __volatile__(
+		"1:	mov	(%0), %%eax\n"	/* read current value */
+		"	test	%%eax, %%eax\n"	/* check if locked */
+		"	jz	3f\n"		/* if not locked, try get it */
+		"2:	pause\n"		/* is locked, hint spinning */
+		"	jmp	1b\n"		/* retry */
+		"3:	lock; cmpxchg %1, (%0)\n" /* try to acquire spinlock */
+		"	jnz	2b\n"		/* if unsuccessful, retry */
+		:
+		: "r" (&lock->lock), "r" (locked)
+		: "eax");
+}
+
+/**
+ * Release previously acquired lock.
+ * @param [in/out] lock Pointer to spinlock
+ */
+static inline void ukarch_spin_unlock(__spinlock *lock)
+{
+	UK_WRITE_ONCE(lock->lock, 0);
+}
+
+/**
+ * Try to acquire lock. If the lock is already acquired (busy), this function
+ * returns instead of spinning.
+ * @param [in/out] lock Pointer to spinlock
+ * @return A non-zero value if lock was acquired, 0 otherwise
+ */
+static inline int ukarch_spin_trylock(__spinlock *lock)
+{
+	register int r = 0, locked = 1;
+
+	__asm__ __volatile__(
+		"	mov	(%1), %%eax\n"	/* read current value */
+		"	test	%%eax, %%eax\n"	/* bail out if locked */
+		"	jnz	1f\n"
+		"	lock; cmpxchg %2, (%1)\n" /* try to acquire spinlock */
+		"	cmove	%2, %0\n"	/* store if successful */
+		"1:\n"
+		: "+&r" (r)
+		: "r" (&lock->lock), "r" (locked)
+		: "eax");
+
+	return r;
```

Comment:
This actually returns zero on successful acquisition, but the documentation states otherwise. Btw, I found this interface inconsistent with `is_locked()`. Maybe we should leave the return value as is and change the interface instead?